### DISCUSSION TITLE
rpc - return the  inner join results as an array in an object

### DISCUSSION THREAD

**Discussion by rodbs** [supabase/supabase#5382] [Questions] [answered]
When running an  inner join using the supabase client it returns a single row  with he join condition in an object: 

```
const { data, error } = await supabase
    .from('order')
    .select('*, order_lines!inner(*)')
    .eq('order.id', '123')
```
it return something like: 
`` [{id:'123', order_lines: [{..}]``

If I want to create a function using this sql, by default it returns as many rows as order_lines.
``` 
select order.*, order_lines.*
  from order 
  inner join order_line
  on order.id = order_line.order_id
```

Is it possible to get one line with all order_lines in a array? Do I need to pre-process it using PL/SQL?

**Comment by steve-chavez** [comment]
> If I want to create a function using this sql, by default it returns as many rows as order_lines.
> Is it possible to get one line with all order_lines in a array

Do you mean you want to get the same json output in SQL as the one that the JS client generates?

**Comment by rodbs** [comment, marked as answer]
I think I've already found what I was looking for :): json_agg
```
select order.id, json_agg(order_lines)
  from order 
  inner join order_line
  on order.id = order_line.order_id
  group by order.id
```

There is no much info about this json functions on supabase's rpc docs. It would be nice to add more examples about json's capabilities on postgresSQL and supabase.

**Reply by nilsi** [reply]
@rodbs how did you create the function? I have this SQL that works but I struggle to create a working RPC function. The function's return type does not match the SQL return since it now has one extra column compared to my listings table.

Here's the SQL I'm trying to make a function out of.

```
SELECT
  l.*,
  json_agg(json_build_object(
    'price', la.price,
    'date', la.date,
    'status', la.status
  )) AS listings_availability
FROM
  listings l
JOIN
  listings_availability la ON la.listing_id = l.id
WHERE
  la.status = 'Available'
  AND la.date BETWEEN '2024-01-01' AND '2024-01-02'
GROUP BY
  l.id
HAVING
  count(*) = ('2024-01-02'::date - '2024-01-01'::date) + 1
ORDER BY
  l.id;
```
